File: stations/_archive/hook-observatory-py/handlers/annotate_insight_hook.py
"""
Annotate Insight PostToolUse handler -- realtime knowledge annotation post-processing.

Features:
  - Listens for mcp__memvault__annotate_insight tool call completion events
  - Optional: calls LiteLLM (qwen3.5-flash) to suggest additional tags for the insight
  - Gracefully skips when LiteLLM is unavailable (does not block main flow)

Design principles:
  - Fire-and-forget: all enrichment logic runs in background, no hook response delay
  - Fail-safe: any exception silently degrades, does not affect annotate_insight main function
"""

from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request

from .base import ALLOW, HOME, HookResult, run_background

# LiteLLM proxy (optional enrichment), gracefully skipped when unavailable
LITELLM_API = os.environ.get("LITELLM_API", "http://127.0.0.1:4000/v1")
LITELLM_MODEL = os.environ.get("ANNOTATE_TAG_MODEL", "qwen3.5-flash")
LITELLM_TIMEOUT = 5  # seconds; fail fast, do not slow down main flow

# tool_name pattern for annotate_insight under mcpproxy
# Claude calls MCP tools with format: mcp__<server>__<tool>
_TOOL_PATTERN = "mcp__memvault__annotate_insight"


def _suggest_tags_via_litellm(insight: str, existing_tags: list[str]) -> list[str]:
    """Call LiteLLM to suggest supplemental tags for the insight (max 3).

    Returns empty list on failure.
    """
    try:
        existing_str = ", ".join(existing_tags) if existing_tags else "(none)"
        prompt = (
            "Given the following insight, suggest 2-3 short tag keywords "
            "(in Traditional Chinese or English) "
            "that would help categorize it for future retrieval. "
            "Return ONLY a JSON array of strings, no explanation.\n\n"
            f"Existing tags: {existing_str}\n"
            f"Insight: {insight[:300]}"
        )

        body = json.dumps(
            {
                "model": LITELLM_MODEL,
                "messages": [{"role": "user", "content": prompt}],
                "max_tokens": 60,
                "temperature": 0.3,
            }
        ).encode()

        req = urllib.request.Request(  # noqa: S310
            f"{LITELLM_API}/chat/completions",
            data=body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )

        with urllib.request.urlopen(req, timeout=LITELLM_TIMEOUT) as resp:  # noqa: S310
            data = json.loads(resp.read().decode())

        raw_content = data["choices"][0]["message"]["content"].strip()

        # parse JSON array from response
        if raw_content.startswith("["):
            suggested = json.loads(raw_content)
            if isinstance(suggested, list):
                # filter out existing tags, take at most 3
                new_tags = [
                    str(t).strip() for t in suggested if str(t).strip() not in existing_tags
                ]
                return new_tags[:3]

    except (urllib.error.URLError, TimeoutError, json.JSONDecodeError, KeyError, IndexError):
        # LiteLLM unavailable or unexpected response format -- silently skip
        return []
    except Exception:
        # unexpected error -- silently degrade, do not block main flow
        return []

    return []


def _enhance_block(block_id: str, insight: str, existing_tags: list[str]) -> None:
    """Background worker: call LiteLLM to supplement tags, then update the memvault block."""
    sys.path.insert(0, os.path.join(HOME, "workshop", "libs", "python", "src"))

    try:
        from sdk_client.clients.memvault import MemvaultClient
    except ImportError:
        return

    # get suggested tags
    new_tags = _suggest_tags_via_litellm(insight, existing_tags)
    if not new_tags:
        return  # no new tags, skip update

    # merge and deduplicate
    merged_tags = list(dict.fromkeys(existing_tags + new_tags))

    try:
        client = MemvaultClient()
        client.update_block(block_id, tags=merged_tags)
    except Exception:
        # update failure silently degrades -- do not block background process
        return


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """PostToolUse handler: annotate_insight completion → background tag enrichment."""
    # only process annotate_insight tool completion events
    if tool_name != _TOOL_PATTERN:
        return ALLOW

    # parse tool output to get block_id
    try:
        data = json.loads(raw_input) if isinstance(raw_input, str) else raw_input
    except (json.JSONDecodeError, TypeError):
        return ALLOW

    # PostToolUse tool_response field contains tool output
    tool_response = data.get("tool_response", "") or ""
    insight = (data.get("tool_input") or tool_input).get("insight", "")
    existing_tags = [
        *((data.get("tool_input") or tool_input).get("tags") or []),
        "realtime-annotation",
    ]

    # parse Block ID from output text (format: "Block ID: <uuid>")
    block_id = ""
    for line in str(tool_response).split("\n"):
        if line.startswith("Block ID:"):
            block_id = line.split(":", 1)[-1].strip()
            break

    # skip enrichment if no valid block_id or insight content
    if not block_id or not insight:
        return ALLOW

    # fire-and-forget: run LiteLLM tag enrichment in background
    python = os.path.join(HOME, ".local", "bin", "python3")
    script = os.path.abspath(__file__)

    run_background(
        [
            python,
            script,
            "--block-id",
            block_id,
            "--insight",
            insight[:300],
            "--tags",
            json.dumps(existing_tags),
        ],
        cwd=os.path.join(HOME, "workshop"),
    )

    return ALLOW


# ---------------------------------------------------------------------------
# Worker mode (called by background process)
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Annotate insight tag enhancer")
    parser.add_argument("--block-id", required=True)
    parser.add_argument("--insight", required=True)
    parser.add_argument("--tags", default="[]")
    args = parser.parse_args()

    try:
        existing = json.loads(args.tags)
    except (json.JSONDecodeError, TypeError):
        existing = []

    _enhance_block(args.block_id, args.insight, existing)

File: internal/handlers/annotate_insight_hook.go
package handlers

// annotate_insight_hook.go — Go port of handlers/annotate_insight_hook.py
//
// PostToolUse / mcp__memvault__annotate_insight:
//   Fires a background goroutine that calls LiteLLM to suggest additional
//   tags for the annotated insight, then PATCHes the memvault block.
//
// Design:
//   - Fire-and-forget: all enrichment runs in a goroutine, handler returns Allow() immediately.
//   - Fail-safe: any error silently degrades, main flow is unaffected.

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/joneshong/hook-observatory/internal/clients"
	"github.com/joneshong/hook-observatory/internal/core"
)

const (
	annotateToolPattern = "mcp__memvault__annotate_insight"
	annotateTagModel    = "qwen3.5-flash"
	annotateMaxTokens   = 60
	annotateTemp        = 0.3
)

// NOTE: Not registered. Python handlers/__init__.py does not import or register
// annotate_insight_hook — it's inactive in production. Keeping the
// implementation for future opt-in, but no core.Register() call to preserve
// parity.
func init() {}

func annotateInsightHandle(_, toolName string, toolInput map[string]any, rawInput string) core.HookResult {
	if toolName != annotateToolPattern {
		return core.Allow()
	}

	// Parse raw input to get tool_response + tool_input fields
	var data map[string]any
	if err := json.Unmarshal([]byte(rawInput), &data); err != nil {
		return core.Allow()
	}

	toolResponse, _ := data["tool_response"].(string)

	// Resolve tool_input: prefer root-level tool_input from JSON, fallback to passed map
	resolvedInput, _ := data["tool_input"].(map[string]any)
	if resolvedInput == nil {
		resolvedInput = toolInput
	}

	insight, _ := resolvedInput["insight"].(string)

	// Collect existing tags
	var existingTags []string
	if tagsRaw, ok := resolvedInput["tags"]; ok {
		switch v := tagsRaw.(type) {
		case []interface{}:
			for _, t := range v {
				if s, ok := t.(string); ok {
					existingTags = append(existingTags, s)
				}
			}
		case []string:
			existingTags = v
		}
	}
	existingTags = append(existingTags, "realtime-annotation")

	// Parse Block ID from tool_response (format: "Block ID: <uuid>")
	blockID := ""
	for _, line := range strings.Split(toolResponse, "\n") {
		if strings.HasPrefix(line, "Block ID:") {
			blockID = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
			break
		}
	}

	if blockID == "" || insight == "" {
		return core.Allow()
	}

	// Fire-and-forget: enrich tags in background goroutine
	go annotateEnrichBackground(blockID, insight, existingTags)

	return core.Allow()
}

// annotateEnrichBackground calls LiteLLM, merges tags, PATCHes memvault block.
// Runs in a goroutine — all errors are silent.
func annotateEnrichBackground(blockID, insight string, existingTags []string) {
	newTags := annotateSuggestTags(insight, existingTags)
	if len(newTags) == 0 {
		return
	}

	// Merge + deduplicate
	seen := make(map[string]struct{}, len(existingTags)+len(newTags))
	merged := make([]string, 0, len(existingTags)+len(newTags))
	for _, t := range existingTags {
		if _, ok := seen[t]; !ok {
			seen[t] = struct{}{}
			merged = append(merged, t)
		}
	}
	for _, t := range newTags {
		if _, ok := seen[t]; !ok {
			seen[t] = struct{}{}
			merged = append(merged, t)
		}
	}

	mc := clients.NewMemvaultClient()
	_, _ = mc.UpdateBlock(blockID, map[string]any{"tags": merged})
}

// annotateSuggestTags calls LiteLLM and returns up to 3 new tag suggestions.
func annotateSuggestTags(insight string, existingTags []string) []string {
	litellmModel := os.Getenv("ANNOTATE_TAG_MODEL")
	if litellmModel == "" {
		litellmModel = annotateTagModel
	}

	existingStr := "(none)"
	if len(existingTags) > 0 {
		existingStr = strings.Join(existingTags, ", ")
	}

	truncated := insight
	if len(truncated) > 300 {
		truncated = truncated[:300]
	}

	prompt := fmt.Sprintf(
		"Given the following insight, suggest 2-3 short tag keywords "+
			"(in Traditional Chinese or English) "+
			"that would help categorize it for future retrieval. "+
			"Return ONLY a JSON array of strings, no explanation.\n\n"+
			"Existing tags: %s\n"+
			"Insight: %s",
		existingStr, truncated,
	)

	content, err := clients.LiteLLMComplete(
		litellmModel,
		[]clients.LiteLLMMessage{{Role: "user", Content: prompt}},
		annotateMaxTokens,
		annotateTemp,
	)
	if err != nil {
		return nil
	}

	raw := strings.TrimSpace(content)
	if !strings.HasPrefix(raw, "[") {
		return nil
	}

	var suggested []interface{}
	if err := json.Unmarshal([]byte(raw), &suggested); err != nil {
		return nil
	}

	// Filter existing, take max 3
	existingSet := make(map[string]struct{}, len(existingTags))
	for _, t := range existingTags {
		existingSet[t] = struct{}{}
	}

	var result []string
	for _, item := range suggested {
		s, ok := item.(string)
		if !ok {
			continue
		}
		s = strings.TrimSpace(s)
		if s == "" {
			continue
		}
		if _, exists := existingSet[s]; !exists {
			result = append(result, s)
		}
		if len(result) >= 3 {
			break
		}
	}
	return result
}
